File: bindings/py/v2/my_lib/network.py
from my_lib._generated import loader
from componentize_py_types import Err

def fetch_and_format(url: str) -> str:
    """
    Fetches the content from a URL via host delegation and formats it in Go WASM.
    """
    res = loader.call("ouvrage:lab-wasi-demo/network", "fetch-and-format", url)
    if res.tag == "err":
        raise Exception(res.payload)
    return res.payload

File: bindings/my_lib/wasm/v2/network.go
//go:build tinygo
package main

import (
	"gitlab.com/ouvrage-systems/labs/wasi-polyglot-bindings-reference/bindings/my_lib/wasm/gen"
)

// Implement exported network interface
type networkImpl struct{}

func (n networkImpl) FetchAndFormat(url string) wasi_polyglot_reference.Result[string, string] {
	// Call the imported host-http function (delegating networking task to the host)
	res := wasi_polyglot_reference.OuvrageLabWasiDemoHostHttpFetchUrl(url)
	if res.IsErr() {
		return wasi_polyglot_reference.Err[string, string]("Go WASM error fetching URL: " + res.UnwrapErr())
	}
	
	val := res.Unwrap()
	formatted := "Go WASM Component formatted: " + val
	return wasi_polyglot_reference.Ok[string, string](formatted)
}
